### ISSUE TITLE
Removing the GoldenEye TLB hack (Pure Interpreter)

### ISSUE THREAD

**Issue by loganmc10** [mupen64plus/mupen64plus-core#447]
I'm referring to this bit of code:

https://github.com/mupen64plus/mupen64plus-core/blob/master/src/device/r4300/tlb.c#L125-L150

GoldenEye actually works with no issues without that code using the Cacher Interpreter or the dynarec. It exits using the Pure Interpreter with this:

```
Core Error: reserved opcode: 8005DD90:1
```
Which it gets to from here:
https://github.com/mupen64plus/mupen64plus-core/blob/master/src/device/r4300/pure_interp.c#L337

@bsmiles32 I'm wondering if you know why this is? It seems strange to me that the Pure Interpreter would arrive at an opcode that the cached interpreter wouldn't

**Comment by bsmiles32** [comment]
This is new to me. I did try some time ago to get rid of this hack, but without it the game refused to boot (but I don't remember which emumode I used). I was more of the impression that the root cause was that PAGEMASK was not honored as it returns (tlb_lookup & ~0xfff) | (address & 0xfff) in the general case and returns (some_base_address + (address & 0xffffff). But I didn't get that far with the investigation, and got distracted by other parts of the core.
But, having the pure_interpreter behaving differently than other emumode is worrying... it's a bug ! Maybe bisecting when this different behavior emerged could help.

**Comment by loganmc10** [comment]
I remember coming across this a long time ago when I was trying it out, I think it's worked without the hack in the cached interpreter and dynarec for a long time. I'm not sure about the new dynarec, which employs the same hack inside its code

**Comment by loganmc10** [comment]
I also tested the x86 new dynarec with the ```tlb_hacks()``` function removed, as well as the bits from ```virtual_to_physical_address()```, and it still works as normal, so it seems like it's just the Pure Interpreter that currently needs that hack to function.

**Comment by loganmc10** [comment]
@bsmiles32 I think the issue relates to how fast_mem_access() works. If an address can't be found via the TLB, virtual_to_physical_address() returns 0. fast_mem_access() then returns mem_base + address (0), and then the Pure Interpreter uses that address to interpret the Opcode.

That doesn't seem right to me, this means that the interpreter will try and execute the opcode found at mem_base + 0 if the TLB can't find the address. I assume it's executing the wrong code, which is causing the game to stop.

I tested this theory and it turned out to be correct. If a TLB exception occurred, InterpretOpcode shouldn't attempt to execute (I had it just return if a TLB exception occurred in fast_mem_access()), this allows GoldenEye to work with the Pure Interpreter without the TLB hack.

I'm not sure about the best/most elegant way to implement this. I did it like:

```
uint32_t *fast_mem_access(struct r4300_core* r4300, uint32_t address , unsigned char* tlb_exp)
 {
     /* This code is performance critical, specially on pure interpreter mode.
      * Removing error checking saves some time, but the emulator may crash. */
 
+    *tlb_exp = 0;
     if ((address & UINT32_C(0xc0000000)) != UINT32_C(0x80000000)) {
         address = virtual_to_physical_address(r4300, address, 2);
+        if (address == 0) *tlb_exp = 1;
     }
....
```

But I'm sure there is a better way.
